modules/text_server_adv/text_server_adv.cpp

```diff
@@ -1728,95 +1768,63 @@ void TextServerAdvanced::shaped_text_overrun_trim_to_width(RID p_shaped_line, fl
 
-					if (above_min_char_treshold && width - ellipsis_advance <= p_width) {
+					if (add_ellipsis && (above_min_char_treshold || enforce_ellipsis) && width - ellipsis_width <= p_width) {
 						ellipsis_pos = trim_pos;
 					}
 					break;
 				}
 			}
 		}
 		if (is_rtl) {
-			width -= sd_glyphs[i].advance;
+			width -= sd_glyphs[i].advance * sd_glyphs[i].repeat;
 		}
 	}
 
+	sd->overrun_info.trim_pos = trim_pos;
+	sd->overrun_info.ellipsis_pos = ellipsis_pos;
+	if (trim_pos == 0 && enforce_ellipsis && add_ellipsis) {
+		sd->overrun_info.ellipsis_pos = 0;
+	}
+
 	if ((trim_pos >= 0 && sd->width > p_width) || enforce_ellipsis) {
-		int added_glyphs = 0;
 		if (add_ellipsis && (ellipsis_pos > 0 || enforce_ellipsis)) {
 			// Insert an additional space when cutting word bound for aesthetics.
 			if (cut_per_word && (ellipsis_pos > 0)) {
 				TextServer::Glyph gl;
-				gl.start = sd_glyphs[ellipsis_pos].start;
-				gl.end = sd_glyphs[ellipsis_pos].end;
 				gl.count = 1;
 				gl.advance = whitespace_adv.x;
 				gl.index = whitespace_gl_idx;
 				gl.font_rid = last_gl_font_rid;
 				gl.font_size = last_gl_font_size;
 				gl.flags = GRAPHEME_IS_SPACE | GRAPHEME_IS_BREAK_SOFT | GRAPHEME_IS_VIRTUAL | (is_rtl ? GRAPHEME_IS_RTL : 0);
 
-				// Optimized glyph insertion by replacing a glyph whenever possible.
-				int glyph_idx = trim_pos + ((is_rtl) ? (-added_glyphs - 1) : added_glyphs);
-				if (is_rtl) {
-					if (glyph_idx < 0) {
-						sd->glyphs.insert(0, gl);
-					} else {
-						sd->glyphs.set(glyph_idx, gl);
-					}
-				} else {
-					if (glyph_idx > (sd_size - 1)) {
-						sd->glyphs.append(gl);
-					} else {
-						sd->glyphs.set(glyph_idx, gl);
-					}
-				}
-				added_glyphs++;
+				sd->overrun_info.ellipsis_glyph_buf.append(gl);
 			}
 			// Add ellipsis dots.
-			for (int d = 0; d < 3; d++) {
-				TextServer::Glyph gl;
-				gl.start = sd_glyphs[ellipsis_pos].start;
-				gl.end = sd_glyphs[ellipsis_pos].end;
-				gl.count = 1;
-				gl.advance = dot_adv.x;
-				gl.index = dot_gl_idx;
-				gl.font_rid = last_gl_font_rid;
-				gl.font_size = last_gl_font_size;
-				gl.flags = GRAPHEME_IS_PUNCTUATION | GRAPHEME_IS_VIRTUAL | (is_rtl ? GRAPHEME_IS_RTL : 0);
-
-				// Optimized glyph insertion by replacing a glyph whenever possible.
-				int glyph_idx = trim_pos + ((is_rtl) ? (-added_glyphs - 1) : added_glyphs);
-				if (is_rtl) {
-					if (glyph_idx < 0) {
-						sd->glyphs.insert(0, gl);
-					} else {
-						sd->glyphs.set(glyph_idx, gl);
-					}
-				} else {
-					if (glyph_idx > (sd_size - 1)) {
-						sd->glyphs.append(gl);
-					} else {
-						sd->glyphs.set(glyph_idx, gl);
-					}
-				}
-				added_glyphs++;
-			}
-		}
-
-		// Cut the remaining glyphs off.
-		if (!is_rtl) {
-			sd->glyphs.resize(trim_pos + added_glyphs);
-		} else {
-			if (trim_pos - added_glyphs >= 0) {
-				sd->glyphs = sd->glyphs.subarray(trim_pos - added_glyphs, sd->glyphs.size() - 1);
-			}
+			TextServer::Glyph gl;
+			gl.count = 3;
```

Comment:
```suggestion
			gl.count = 1;
```

`count` should be `1`, it's there to indicate the number of subsequent glyphs that belong to the same grapheme (e.g. when combining diacritic for "Ā" is used, the first glyph in the `A`, `¯` sequence have count = 2).

Reply:
Fixed.